qutebrowser/javascript/history.js

```diff
@@ -0,0 +1,194 @@
+
+            // Add table to page
+            HIST_CONTAINER.appendChild(table);
+        }
+
+        // Find/create tbody
+        var tbody = table.lastChild;
+        if (tbody.tagName !== "TBODY") {
+            tbody = document.createElement("tbody");
+            table.appendChild(tbody);
+        }
+
+        // Create session-separator and new tbody if necessary
+        if (tbody.lastChild !== null && lastTime !== null) {
+            var lastItemDate = new Date(lastTime * 1000);
+            var interval = (lastItemDate.getTime() - date.getTime()) / 1000.00;
+            if (interval > SESSION_CUTOFF) {
+                // Add session-separator
+                var sessionSeparator = document.createElement("td");
+                sessionSeparator.className = "session-separator";
+                sessionSeparator.colSpan = 2;
+                sessionSeparator.innerHTML = "&#167;";
+                table.appendChild(document.createElement("tr"));
+                table.lastChild.appendChild(sessionSeparator);
+
+                // Create new tbody
+                tbody = document.createElement("tbody");
+                table.appendChild(tbody);
+            }
+        }
+
+        return tbody;
+    }
+
+    /**
+     * Given a history item, create and return <tr> for it.
+     *
+     * @param {string} itemUrl - The url for this item.
+     * @param {string} itemTitle - The title for this item.
+     * @param {string} itemTime - The formatted time for this item.
+     * @returns {Element} the completed tr.
+     */
+    function makeHistoryRow(itemUrl, itemTitle, itemTime) {
+        var row = document.createElement("tr");
+
+        var title = document.createElement("td");
+        title.className = "title";
+        var link = document.createElement("a");
+        link.href = itemUrl;
+        link.innerHTML = itemTitle;
+        title.appendChild(link);
+
+        var time = document.createElement("td");
+        time.className = "time";
+        time.innerHTML = itemTime;
+
+        row.appendChild(title);
+        row.appendChild(time);
+
+        return row;
+    }
+
+    /**
+     * Get JSON from given URL.
+     *
+     * @param {string} url - the url to fetch data from.
+     * @param {function} callback - the function to callback with data.
+     * @returns {void}
+     */
+    function getJSON(url, callback) {
+        var xhr = new XMLHttpRequest();
+        xhr.open("GET", url, true);
+        xhr.responseType = "json";
+        xhr.onload = function() {
+            var status = xhr.status;
+            callback(status, xhr.response);
+        };
+        xhr.send();
+    }
+
+    /**
+     * Receive history data from qute://history/data.
+     *
+     * @param {Number} status - The status of the query.
+     * @param {Array} history - History data.
+     * @returns {void}
+     */
+    function receiveHistory(status, history) {
+        if (history === undefined) {
+            return;
```

Comment:
How can history be `undefined`? I guess you should rather check for `null`.
https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/response